sno/diff_output.py

```diff
@@ -283,6 +284,23 @@ def _out(dataset, diff):
 
 
 def _json_row(row, change, pk_field):
+    f = {
+        "geometry": None,
+        "properties": {},
+        "id": f"{change}::{row[pk_field]}",
+    }
+
+    for k in row.keys():
+        v = row[k]
+        if isinstance(v, bytes):
+            f["geometry"] = gpkg.gpkg_geom_to_hex_wkb(v)
```

Comment:
as we've ditched geojson, should we make it a `{field: value}` dict and drop properties/geometry/etc?

Reply:
I was going to get rid of `id`, but it's kind of useful since when you're looking at a patch you don't know what that PK field is.

`geometry` is sort-of the same, although there's two problems:

- at present you _still_ don't know what the geometry field is
- if we moved it from `geometry` into `properties.<fieldname>` you gain the field name but lose the fact that it's a geometry field

For applying patches, the dataset already has type information so I guess the patch doesn't _have_ to. So I'd tend to just put all properties together, including geometry. So remove _all_ of `geometry`, `id` and `properties` and just make the feature a dict of `fieldname: value`

This raises the question of whether patches should include schema information for conflict detection, so if you change the PK field or the geometry field the patch can be rejected. Thoughts?

That all might be worth a separate effort though; this PR could be just about old enough

Reply:
On `id`: I think if we keep `id` then it should cease to be a weird `"U-::2"` thing and should just be the value of the PK field. (This applies to geojson output too). That's a separate PR though.

Reply:
> On id: I think if we keep id then it should cease to be a weird "U-::2" thing and should just be the value of the PK field. (This applies to geojson output too).

`id`'s need to be unique in GeoJSON, so can't be the PK field (because there's a new & old side)?

Reply:
> This raises the question of whether patches should include schema information for conflict detection, so if you change the PK field or the geometry field the patch can be rejected. Thoughts?

I think including the schema hash from the new data format might make sense? That will indicate compatibility, so if the hash doesn't match and there's no schema-change information then you can't really apply it.

Reply:
(but yes, no point doing that until it's implemented)

Reply:
> id's need to be unique in GeoJSON, so can't be the PK field (because there's a new & old side)?
okay. So remove the `id` altogether in other formats, but leave it in the existing form for GeoJSON 👍

Reply:
to clarify, I'm not planning to do that in this PR; we can clean it up in another PR